sdk/python/v1beta1/kubeflow/katib/api/katib_client.py

```diff
@@ -267,13 +268,17 @@ def list_experiments(self, namespace=None):
         katibexp = None
         try:
             katibexp = thread.get(constants.APISERVER_TIMEOUT)
-            result = []
-            for i in katibexp.get("items"):
-                output = {}
-                output["name"] = i.get("metadata", {}).get("name")
-                output["status"] = i.get("status", {}).get(
-                    "conditions", [])[-1].get("type")
-                result.append(output)
+            result = [
+                V1beta1Experiment(
+                    api_version=item.get("apiVersion"),
+                    kind=item.get("kind"),
+                    metadata=item.get("metadata"),
+                    spec=item.get("spec"),
```

Comment:
In that case what will be the type for `experiment.spec` ?
I was thinking maybe we should use [`deserialize`](https://github.com/kubeflow/katib/blob/427ca71006103627230a1532352357c02cd32e7c/sdk/python/v1beta1/kubeflow/katib/api_client.py#L260) method to convert result to the List of objects.
What do you think @anencore94 @gaocegege @johnugeorge ?

Reply:
That would be better! Thanks

Reply:
With new method, I could check V1beta1Experiment object's nested type was filled as I intended.
Here is the test case:
- Python code
```
    resp = client.list_experiments(namespace='jaeyeon-kim')
    print(resp[0])
    print(type(resp[0]))
    print(type(resp[0].metadata))
    print(type(resp[0].spec))
    print(type(resp[0].spec.algorithm))

    resp2 = client.list_trials(name='katib-hpo-test', namespace='jaeyeon-kim')
    print(resp2[0])
    print(type(resp2[0]))
    print(type(resp2[0].spec.objective))
```

- The stdout with some omitted
```
{'api_version': 'kubeflow.org/v1beta1',
 'kind': 'Experiment',
 'metadata': {'annotations': {'kubectl.kubernetes.io/last-applied-configuration': '{"apiVersion":"kubeflow.org/v1beta1","kind":"Experiment","metadata":{"annotations":{},"name":"katib-hpo-test","namespace":"jaeyeon-kim"},"spec":{"algorithm":{"algorithmName":"tpe"},"maxFailedTrialCount":2,"maxTrialCount":5,"metricsCollectorSpec":{"collector":{"kind":"File"},"source":{"fileSystemPath":{"kind":"File","path":"/var/log/katib/mrx-metrics.log"}}},"objective":{"goal":0.001,"objectiveMetricName":"loss","type":"minimize"},"parallelTrialCount":2,"parameters":[{"feasibleSpace":{"max":"2.0","min":"0.0"},"name":"x","parameterType":"double"},{"feasibleSpace":{"max":"3.0","min":"0.0"},"name":"y","parameterType":"double"},{"feasibleSpace":{"max":"5.0","min":"1.0"},"name":"z","parameterType":"double"}],"trialTemplate":{"failureCondition":"status.conditions.#(type==\\"Failed\\")#|#(status==\\"True\\")#","primaryContainerName":"katib-helper","successCondition":"status.conditions.#
......
......
              'name': 'katib-hpo-test',
              'namespace': 'jaeyeon-kim',
              'owner_references': None,
              'resource_version': '422218026',
              'self_link': '/apis/kubeflow.org/v1beta1/namespaces/jaeyeon-kim/experiments/katib-hpo-test',
              'uid': '7f930fc1-9de6-474c-9417-5a41c9339498'},
 'spec': {'algorithm': {'algorithm_name': 'tpe', 'algorithm_settings': None},
          'early_stopping': None,
          'max_failed_trial_count': 2,
          'max_trial_count': 5,
          'metrics_collector_spec': {'collector': {'custom_collector': None,
                                                   'kind': 'File'},
                                     'source': {'file_system_path': {'kind': 'File',
                                                                     'path': '/var/log/katib/mrx-metrics.log'},
                                                'filter': None,
                                                'http_get': None}},
          'nas_config': None,
          'objective': {'additional_metric_names': None,
                        'goal': 0.001,
                        'metric_strategies': [{'name': 'loss', 'value': 'min'}],
                        'objective_metric_name': 'loss',
                        'type': 'minimize'},
          'parallel_trial_count': 2,
          'parameters': [{'feasible_space': {'list': None,
                                             'max': '2.0',
                                             'min': '0.0',
                                             'step': None},
                          'name': 'x',
                          'parameter_type': 'double'},
                         {'feasible_space': {'list': None,
                                             'max': '3.0',
                                             'min': '0.0',
                                             'step': None},
                          'name': 'y',
                          'parameter_type': 'double'},
                         {'feasible_space': {'list': None,
                                             'max': '5.0',
                                             'min': '1.0',
                                             'step': None},
                          'name': 'z',
                          'parameter_type': 'double'}],
          'resume_policy': 'LongRunning',
          'trial_template': {'config_map': None,
                             'failure_condition': 'status.conditions.#(type=="Failed")#|#(status=="True")#',
                             'primary_container_name': 'katib-helper',
                             'primary_pod_labels': None,
                             'retain': None,
......
......
                                                                                           'imagePullPolicy': 'IfNotPresent',
                                                                                           'name': 'katib-helper'}],
                                                                           'imagePullSecrets': [{'name': 'regcred-ghcr'}],
                                                                           'restartPolicy': 'Never',
                                                                           'serviceAccountName': 'katib-account'}}}}}},
 'status': {'completion_time': datetime.datetime(2021, 8, 21, 14, 9, 27, tzinfo=tzutc()),
            'conditions': [{'last_transition_time': datetime.datetime(2021, 8, 21, 14, 6, 5, tzinfo=tzutc()),
                            'last_update_time': datetime.datetime(2021, 8, 21, 14, 6, 5, tzinfo=tzutc()),
                            'message': 'Experiment is created',
                            'reason': 'ExperimentCreated',
                            'status': 'True',
                            'type': 'Created'},
                           {'last_transition_time': datetime.datetime(2021, 8, 21, 14, 9, 27, tzinfo=tzutc()),
                            'last_update_time': datetime.datetime(2021, 8, 21, 14, 9, 27, tzinfo=tzutc()),
                            'message': 'Experiment is running',
                            'reason': 'ExperimentRunning',
                            'status': 'False',
                            'type': 'Running'},
                           {'last_transition_time': datetime.datetime(2021, 8, 21, 14, 9, 27, tzinfo=tzutc()),
                            'last_update_time': datetime.datetime(2021, 8, 21, 14, 9, 27, tzinfo=tzutc()),
                            'message': 'Experiment has succeeded because max '
                                       'trial count has reached',
                            'reason': 'ExperimentMaxTrialsReached',
                            'status': 'True',
                            'type': 'Succeeded'}],
            'current_optimal_trial': {'best_trial_name': 'katib-hpo-test-g4wkct7w',
                                      'observation': {'metrics': [{'latest': '0.16119065270234822',
                                                                   'max': '0.16119065270234822',
                                                                   'min': '0.16119065270234822',
                                                                   'name': 'loss'}]},
                                      'parameter_assignments': [{'name': 'x',
                                                                 'value': '0.8376333862283576'},
                                                                {'name': 'y',
                                                                 'value': '2.3292919169041912'},
                                                                {'name': 'z',
                                                                 'value': '3.0852625163721057'}]},
            'early_stopped_trial_list': None,
            'failed_trial_list': None,
            'killed_trial_list': None,
            'last_reconcile_time': None,
            'pending_trial_list': None,
            'running_trial_list': None,
            'start_time': datetime.datetime(2021, 8, 21, 14, 6, 5, tzinfo=tzutc()),
            'succeeded_trial_list': ['katib-hpo-test-b67wvc6l',
                                     'katib-hpo-test-g4wkct7w',
                                     'katib-hpo-test-nfpb5k75',
                                     'katib-hpo-test-znjhw7rt',
                                     'katib-hpo-test-xphcc7hm'],
            'trials': 5,
            'trials_early_stopped': None,
            'trials_failed': None,
            'trials_killed': None,
            'trials_pending': None,
            'trials_running': None,
            'trials_succeeded': 5}}
<class 'kubeflow.katib.models.v1beta1_experiment.V1beta1Experiment'>
<class 'kubernetes.client.models.v1_object_meta.V1ObjectMeta'>
<class 'kubeflow.katib.models.v1beta1_experiment_spec.V1beta1ExperimentSpec'>
<class 'kubeflow.katib.models.v1beta1_algorithm_spec.V1beta1AlgorithmSpec'>

---------------------------------------------------------------------------------------------
{'api_version': 'kubeflow.org/v1beta1',
 'kind': 'Trial',
 'metadata': {'annotations': None,
              'cluster_name': None,
              'creation_timestamp': datetime.datetime(2021, 8, 21, 14, 5, 16, tzinfo=tzutc()),
              'deletion_grace_period_seconds': None,
              'deletion_timestamp': None,
              'finalizers': ['clean-metrics-in-db'],
              'generate_name': None,
              'generation': 1,
              'labels': {'experiment': 'katib-hpo-test-2'},
              'managed_fields': [{'api_version': 'kubeflow.org/v1beta1',
                                  'fields_type': 'FieldsV1',
                                  'fields_v1': {'f:metadata': {'f:finalizers': {},
                                                               'f:labels': {'.': {},
                                                                            'f:experiment': {}},
                                                               'f:ownerReferences': {}},
                                                'f:spec': {'.': {},
                                                           'f:failureCondition': {},
                                                           'f:metricsCollector': {'.': {},
                                                                                  'f:collector': {'.': {},
                                                                                                  'f:kind': {}},
                                                                                  'f:source': {'.': {},
                                                                                               'f:fileSystemPath': {'.': {},
                                                                                                                    'f:kind': {},
                                                                                                                    'f:path': {}}}},
                                                           'f:objective': {'.': {},
                                                                           'f:goal': {},
                                                                           'f:metricStrategies': {},
                                                                           'f:objectiveMetricName': {},
                                                                           'f:type': {}},
                                                           'f:parameterAssignments': {},
                                                           'f:primaryContainerName': {},
                                                           'f:runSpec': {'.': {},
                                                                         'f:apiVersion': {},
                                                                         'f:kind': {},
                                                                         'f:metadata': {'.': {},
                                                                                        'f:creationTimestamp': {},
                                                                                        'f:labels': {'.': {},
                                                                                                     'f:app.kubernetes.io/part-of': {},
                                                                                                     'f:controller-uid': {},
                                                                                                     'f:job-name': {}},
                                                                                        'f:managedFields': {},
                                                                                        'f:name': {},
                                                                                        'f:namespace': {},
                                                                                        'f:ownerReferences': {},
                                                                                        'f:resourceVersion': {},
                                                                                        'f:selfLink': {},
                                                                                        'f:uid': {}},
                                                                         'f:spec': {'.': {},
                                                                                    'f:backoffLimit': {},
                                                                                    'f:completions': {},
                                                                                    'f:parallelism': {},
                                                                                    'f:selector': {'.': {},
                                                                                                   'f:matchLabels': {'.': {},
                                                                                                                     'f:controller-uid': {}}},
                                                                                    'f:template': {'.': {},
                                                                                                   'f:metadata': {'.': {},
                                                                                                                  'f:annotations': {'.': {},
                                                                                                                                    'f:sidecar.istio.io/inject': {}},
                                                                                                                  'f:creationTimestamp': {},
                                                                                                                  'f:labels': {'.': {},
                                                                                                                               'f:app.kubernetes.io/part-of': {},
                                                                                                                               'f:controller-uid': {},
                                                                                                                               'f:job-name': {}}},
                                                                                                   'f:spec': {'.': {},
                                                                                                              'f:containers': {},
                                                                                                              'f:dnsPolicy': {},
                                                                                                              'f:imagePullSecrets': {},
                                                                                                              'f:restartPolicy': {},
                                                                                                              'f:schedulerName': {},
                                                                                                              'f:securityContext': {},
                                                                                                              'f:serviceAccount': {},
                                                                                                              'f:serviceAccountName': {},
                                                                                                              'f:terminationGracePeriodSeconds': {}}}},
                                                                         'f:status': {}},
                                                           'f:successCondition': {}},
                                                'f:status': {'.': {},
                                                             'f:completionTime': {},
                                                             'f:conditions': {},
                                                             'f:observation': {'.': {},
                                                                               'f:metrics': {}},
                                                             'f:startTime': {}}},
                                  'manager': 'katib-controller',
                                  'operation': 'Update',
                                  'time': datetime.datetime(2021, 8, 21, 14, 6, 22, tzinfo=tzutc())}],
              'name': 'katib-hpo-test-2-b9pf6btg',
              'namespace': 'jaeyeon-kim',
              'owner_references': [{'api_version': 'kubeflow.org/v1beta1',
                                    'block_owner_deletion': True,
                                    'controller': True,
                                    'kind': 'Experiment',
                                    'name': 'katib-hpo-test-2',
                                    'uid': '83155259-8944-42f2-a288-0afc4b2dfb68'}],
              'resource_version': '422207197',
              'self_link': '/apis/kubeflow.org/v1beta1/namespaces/jaeyeon-kim/trials/katib-hpo-test-2-b9pf6btg',
              'uid': '18e5ac5c-ca24-46d7-9acb-6a886bf2f09a'},
 'spec': {'early_stopping_rules': None,
          'failure_condition': 'status.conditions.#(type=="Failed")#|#(status=="True")#',
          'metrics_collector': {'collector': {'custom_collector': None,
                                              'kind': 'File'},
                                'source': {'file_system_path': {'kind': 'File',
                                                                'path': '/var/log/katib/mrx-metrics.log'},
                                           'filter': None,
                                           'http_get': None}},
          'objective': {'additional_metric_names': None,
                        'goal': 0.001,
                        'metric_strategies': [{'name': 'loss', 'value': 'min'}],
                        'objective_metric_name': 'loss',
                        'type': 'minimize'},
          'parameter_assignments': [{'name': 'x',
                                     'value': '0.8240284763063495'},
                                    {'name': 'y',
                                     'value': '1.8777500963698879'},
                                    {'name': 'z',
                                     'value': '4.132651796238958'}],
          'primary_container_name': 'katib-helper',
          'primary_pod_labels': None,
          'retain_run': None,
          'run_spec': {'apiVersion': 'batch/v1',
                       'kind': 'Job',
                       'metadata': {'name': 'katib-hpo-test-2-b9pf6btg',
                                    'namespace': 'jaeyeon-kim'},
                       'spec': {'template': {'metadata': {'annotations': {'sidecar.istio.io/inject': 'true'},
                                                          'labels': {'app.kubernetes.io/part-of': 'katib-experiment'}},
                                             'spec': {'containers': [{'command': ['until '
                                                                                  '$(curl '
                                                                                  '--output '
                                                                                  '/dev/null '
                                                                                  '--silent '
                                                                                  '--head '
                                                                                  '--fail '
                                                                                  'http://localhost:15021/healthz/ready); '
                                                                                  'do '
                                                                                  'echo '
                                                                                  '. '
                                                                                  '; '
                                                                                  'sleep '
                                                                                  '5; '
                                                                                  'done; '
                                                                                  'kfp '
                                                                                  'run '
                                                                                  'submit '
                                                                                  '-t '
                                                                                  '300 '
                                                                                  '-e '
                                                                                  '1442bc8a-bc0e-482d-8658-2553abf7257e '
                                                                                  '-p '
                                                                                  '780544e6-40e9-45ab-9987-322988b91edf '
                                                                                  'x=0.8240284763063495 '
                                                                                  'y=1.8777500963698879 '
                                                                                  'z=4.132651796238958 '
                                                                                  '> '
                                                                                  '/tmp/run-submit-result; '
                                                                                  'export '
                                                                                  'RUN_ID=$(awk '
                                                                                  "'/Succeeded/' "
                                                                                  '/tmp/run-submit-result '
                                                                                  '| '
                                                                                  'awk '
                                                                                  '-F '
                                                                                  "'| "
                                                                                  "' "
                                                                                  "'{ "
                                                                                  'print '
                                                                                  '$2 '
                                                                                  "}'); \n"
                                                                                  '    '
                                                                                  '(echo '
                                                                                  '"import '
                                                                                  'kfp";\n'
                                                                                  '     '
                                                                                  'echo '
                                                                                  '"import '
                                                                                  'os";\n'
                                                                                  '     '
                                                                                  'echo '
                                                                                  '"import '
                                                                                  'json";\n'
                                                                                  '     '
                                                                                  'echo '
                                                                                  '"client=kfp.Client()";\n'
                                                                                  '     '
                                                                                  'echo '
                                                                                  '"run_response '
                                                                                  '= '
                                                                                  'client.get_run(os.getenv(\'RUN_ID\'))";\n'
                                                                                  '     '
                                                                                  'echo '
                                                                                  '"for '
                                                                                  'metric '
                                                                                  'in '
                                                                                  'run_response.run.metrics: '
                                                                                  'print(f\'{metric.name}={metric.number_value}\')"\n'
                                                                                  '     '
                                                                                  ') '
                                                                                  '| '
                                                                                  'python '
                                                                                  '>> '
                                                                                  '/var/log/katib/mrx-metrics.log\\\n'
                                                                                  '     '],
                                                                      'image': 'ghcr.io/makinarocks/katib-helper:v1.0.2',
                                                                      'imagePullPolicy': 'IfNotPresent',
                                                                      'name': 'katib-helper'}],
                                                      'imagePullSecrets': [{'name': 'regcred-ghcr'}],
                                                      'restartPolicy': 'Never',
                                                      'serviceAccountName': 'katib-account'}}}},
          'success_condition': 'status.conditions.#(type=="Complete")#|#(status=="True")#'},
 'status': {'completion_time': datetime.datetime(2021, 8, 21, 14, 6, 22, tzinfo=tzutc()),
            'conditions': [{'last_transition_time': datetime.datetime(2021, 8, 21, 14, 5, 16, tzinfo=tzutc()),
                            'last_update_time': datetime.datetime(2021, 8, 21, 14, 5, 16, tzinfo=tzutc()),
                            'message': 'Trial is created',
                            'reason': 'TrialCreated',
                            'status': 'True',
                            'type': 'Created'},
                           {'last_transition_time': datetime.datetime(2021, 8, 21, 14, 6, 22, tzinfo=tzutc()),
                            'last_update_time': datetime.datetime(2021, 8, 21, 14, 6, 22, tzinfo=tzutc()),
                            'message': 'Trial is running',
                            'reason': 'TrialRunning',
                            'status': 'False',
                            'type': 'Running'},
                           {'last_transition_time': datetime.datetime(2021, 8, 21, 14, 6, 22, tzinfo=tzutc()),
                            'last_update_time': datetime.datetime(2021, 8, 21, 14, 6, 22, tzinfo=tzutc()),
                            'message': 'Trial has succeeded',
                            'reason': 'TrialSucceeded',
                            'status': 'True',
                            'type': 'Succeeded'}],
            'last_reconcile_time': None,
            'observation': {'metrics': [{'latest': '1.1150389600458972',
                                         'max': '1.1150389600458972',
                                         'min': '1.1150389600458972',
                                         'name': 'loss'}]},
            'start_time': datetime.datetime(2021, 8, 21, 14, 5, 16, tzinfo=tzutc())}}
<class 'kubeflow.katib.models.v1beta1_trial.V1beta1Trial'>
<class 'kubeflow.katib.models.v1beta1_objective_spec.V1beta1ObjectiveSpec'>

Process finished with exit code 0

```

Reply:
Thank you @anencore94!
> That would be better! Thanks. But I have to add similar method : deserialize_data, to deserialize json, not RESTResponse object from thread.get()

Do you know, what is the type in [`item` here](https://github.com/kubeflow/katib/blob/f0308997475a3787668aedc21b63b37eb270d57e/sdk/python/v1beta1/kubeflow/katib/api/katib_client.py#L274) ?
It is not `RESTResponse` ?

Have you tried to run `deserialize(list[V1beta1Experiment])` ?

Reply:
> Do you know, what is the type in item here ?
It is not RESTResponse ?

Testing with following code in  [item here ](https://github.com/kubeflow/katib/blob/f0308997475a3787668aedc21b63b37eb270d57e/sdk/python/v1beta1/kubeflow/katib/api/katib_client.py#L274)
```
        katibexp = None
        try:
            katibexp = thread.get(constants.APISERVER_TIMEOUT)

            print(type(katibexp))
            for item in katibexp.get("items"):
                print(type(item))
                print(item)
                sys.exit(1)
```

The stdout was as follows:
```
<class 'dict'>
<class 'dict'>
{'apiVersion': 'kubeflow.org/v1beta1', 'kind': 'Experiment', 'metadata': {'annotations': {'kubectl.kubernetes.io/last-applied-configuration': '{"apiVersion":"kubeflow.org/v1beta1","kind":"Experiment","metadata":{"annotations":{},"name":"hpo-quadratic","namespace":"jaeyeon-kim"},"spec":{"algorithm":{"algorithmName":"tpe"},"maxFailedTrialCount":2,"maxTrialCount":3,"metricsCollectorSpec":{"collector":{"kind":"File"},"source":{"fileSystemPath":{"kind":"File","path":"/var/log/katib/mrx-metrics.log"}}},"objective":{"goal":0.5,"objectiveMetricName":"loss","type":"minimize"},"parallelTrialCount":1,"parameters":[{"feasibleSpace":{"max":"1.0","min":"0.0"},"name":"x","parameterType":"double"},{"feasibleSpace":{"max":"3.0","min":"0.0"},"name":"y","parameterType":"double"},{"feasibleSpace":{"max":"65.0","min":"0.0"},"name":"z","parameterType":"double"}],"trialTemplate":{"failureCondition":"status.conditions.#(type==\\"Failed\\")#|#(status==\\"True\\")#","primaryContainerName":"katib-helper","successCondition":"status.conditions.#(type==\\"Complete\\")#|#(status==\\"True\\")#","trialParameters":[{"name":"x","reference":"x"},{"name":"y","reference":"y"},{"name":"z","reference":"z"}],"trialSpec":{"apiVersion":"batch/v1","kind":"Job",
.....
// omitted
}
```

Reply:
Since both instances' type was dict, not RESTResponse, I couldn't use existing [deserialize](https://github.com/kubeflow/katib/blob/427ca71006103627230a1532352357c02cd32e7c/sdk/python/v1beta1/kubeflow/katib/api_client.py#L260) function

Reply:
Yeah, I think it's common issue: https://github.com/kubernetes-client/python/issues/340.
I also tried this: https://github.com/kubernetes-client/python/issues/574#issuecomment-405400414 with `_preload_content=False`, but `list_namespaced_custom_object` can't be converted still.
And it requires `kubernetes.client.models` converter for it as you did.

Maybe we can create our own `deserialize` method in utils: https://github.com/kubeflow/katib/blob/master/sdk/python/v1beta1/kubeflow/katib/utils/utils.py ?

What do you think?

Reply:
Yeap, that makes sense @andreyvelich , Thanks for checking this further :)
Then you mean the original [deserialize](https://github.com/kubeflow/katib/blob/427ca71006103627230a1532352357c02cd32e7c/sdk/python/v1beta1/kubeflow/katib/api_client.py#L260) function should be as before, but we make another method something like `deserialize(data: Union[dict, list, str], data_type: Type[T]) -> T` right ?

Reply:
Yes, we can make another method as you made [here](https://github.com/kubeflow/katib/blob/f0308997475a3787668aedc21b63b37eb270d57e/sdk/python/v1beta1/kubeflow/katib/api_client.py#L286).

The problem with this conversion: https://github.com/kubeflow/katib/blob/f0308997475a3787668aedc21b63b37eb270d57e/sdk/python/v1beta1/kubeflow/katib/api_client.py#L323-L326.

Is there any chance to remove it and make deserialization?
We defined Kubernetes models in Katib models: https://github.com/kubeflow/katib/blob/f0308997475a3787668aedc21b63b37eb270d57e/sdk/python/v1beta1/kubeflow/katib/models/__init__.py#L59-L63.
So it's strange why `getattr` doesn't work.

Reply:
I changed the location of `deserialize` method to utils.py, so I need to copy all existing private methods in api_client.py.

Also, I still needs to check `dir(kubernetes.client.models)` even if they were in `dir(kubeflow.katib.models)` as you mentioned.

Reply:
I think I found the reason why.
There were missing class in katib/sdk/python/v1beta1/kubeflow/katib/models/__init__.py :
```
from kubernetes.client import V1ManagedFieldsEntry
from kubernetes.client import V1OwnerReference
```

However, even though I tried to add them manually, `make generate` automatically removes such lines.
Do I have to specify those import packages at [swagger_config.json](https://github.com/kubeflow/katib/blob/21580f66972640ee60aea0f86d74a4f496ee98c3/hack/gen-python-sdk/swagger_config.json#L5) or somewhere else ?

Reply:
>Do I have to specify those import packages at swagger_config.json or somewhere else ?

Exactly, let's add these import to swagger config.